src/main/java/com/nccgroup/loggerplusplus/exports/CobaltExporter.java

```diff
@@ -158,36 +155,52 @@ private void indexPendingEntries(){
-                }
-                connectFailedCounter = 0;
-            } catch (ConnectException e) {
-                LoggerPlusPlus.callbacks.printError("Connection error, upload failed");
-                connectFailedCounter++;
-                if(connectFailedCounter > 5) {
-                    JOptionPane.showMessageDialog(JOptionPane.getFrameForComponent(LoggerPlusPlus.instance.getLoggerMenu()),
-                            "Cobalt exporter could not connect after 5 attempts. Elastic exporter shutting down...",
-                            "Cobalt Exporter - Connection Failed", JOptionPane.ERROR_MESSAGE);
-                    shutdown();
+            for (List<LogEntry> entries : batchEntries) {
+                try {
+                    LoggerPlusPlus.callbacks.printOutput("Uploading 1 message with " + entries.size() + " log entries...");
+                    post.setEntity(buildRequestBody((ArrayList<LogEntry>) entries));
+                    CloseableHttpResponse response = httpClient.execute(post);
+                    int statusCode = response.getStatusLine().getStatusCode();
+
+                    LoggerPlusPlus.callbacks.printOutput("Upload finished with status code " + statusCode);
+                    if (statusCode >= 400) {
+                        LoggerPlusPlus.callbacks.printOutput(EntityUtils.toString(response.getEntity()));
+                    }
+                    connectFailedCounter = 0;
+                } catch (ConnectException e) {
+                    LoggerPlusPlus.callbacks.printError("Connection error, upload failed");
+                    connectFailedCounter++;
+                    if (connectFailedCounter > 5) {
+                        JOptionPane.showMessageDialog(JOptionPane.getFrameForComponent(LoggerPlusPlus.instance.getLoggerMenu()),
+                                "Cobalt exporter could not connect after 5 attempts. Elastic exporter shutting down...",
+                                "Cobalt Exporter - Connection Failed", JOptionPane.ERROR_MESSAGE);
+                        shutdown();
+                    }
+                } catch (Exception e) {
+                    LoggerPlusPlus.callbacks.printError("Upload failed: " + e.getMessage());
+                } finally {
+                    post.releaseConnection();
                 }
             }
-        }catch (Exception e){
-                LoggerPlusPlus.callbacks.printError("Upload failed: " + ExceptionUtils.getStackTrace(e));
+        } catch (Exception e) {
+            LoggerPlusPlus.callbacks.printError("Upload failed: " + ExceptionUtils.getStackTrace(e));
         }
     }
 
+    private Collection<List<LogEntry>> splitPendingEntries(ArrayList<LogEntry> entriesInBulk) {
+        AtomicInteger counter = new AtomicInteger();
+        return entriesInBulk
+                .stream()
+                .collect(Collectors.groupingBy(it -> counter.getAndIncrement() / UPLOAD_BATCH_SIZE))
```

Comment:
I don't have that much experience with java, but I've never seen division with atomic integer. When reviewing this code I don't know how this behaves in a general case. I believe integer a / b this can be 0 in case a < b.

Reply:
Java integer division will ignore the reminder values.

In our code,
For example lets say we have 5 elements in list and we want to split in batches of max 2 elements.
In the above code, all the elements matching the result of code `counter.getAndIncrement() / UPLOAD_BATCH_SIZE)` will be grouped together.

Example list: ['a', 'b', 'c', 'd', 'e']
UPLOAD_BATCH_SIZE: 2
For each iteration result of `counter.getAndIncrement() / UPLOAD_BATCH_SIZE` will be,
Iteration 1: 0/2 = 0
Iteration 2: 1/2 = 0
Iteration 3: 2/2 = 1
Iteration 4: 3/2 = 1
Iteration 5: 4/2 = 2

So the list will be grouped like this: [['a','b'], ['c','d'], ['e']]

Not sure if it answered your question :)

Reply:
Oh! I see! It returns index of the group. Thank you for clarification.